### PR TITLE
GH-118074: Executors in the COLD_EXITS array are not GC'able

### PR DESCRIPTION
**Pull request by gvanrossum**
Implement a `tp_is_gc` slot that tests for this.

* Issue: gh-118074

### CODE REVIEW

**Comment by gvanrossum** [comment]
@adoxalim Can you verify that this fix works for you? (Thanks again for the bug report.)

File: Python/optimizer.c [hunk 1]
```diff
@@ -394,6 +394,15 @@ executor_traverse(PyObject *o, visitproc visit, void *arg)
     return 0;
 }
 
+static int
+executor_is_gc(PyObject *o)
+{
+    if ((PyObject *)&COLD_EXITS[0] <= o && o < (PyObject *)&COLD_EXITS[UOP_MAX_TRACE_LENGTH]) {
```
Comment (mdboom) [inline Python/optimizer.c]: The address sanitizer warning here seems plausibly correct:

```
array subscript 800 is above array bounds of ‘_PyExecutorObject[200]’ [-Warray-bounds]
```

`COLD_EXITS` has a length of `UOP_MAX_TRACE_LENGTH/4`.  If there's some other reason this is correct, maybe add a comment / a way to suppress the warning?

Comment (gvanrossum) [reply Python/optimizer.c]: Oh, you're right. That line was written by Copilot and it looked so plausible but was so wrong. :-(

**Comment by gvanrossum** [comment]
The emulated Linux (ARM64) test failures are known, caused by the the QEMU emulation. I'm merging now.

File: Python/optimizer.c [hunk 2]
```diff
@@ -394,6 +394,15 @@ executor_traverse(PyObject *o, visitproc visit, void *arg)
     return 0;
 }
 
+static int
+executor_is_gc(PyObject *o)
+{
+    if ((PyObject *)&COLD_EXITS[0] <= o && o < (PyObject *)&COLD_EXITS[COLD_EXIT_COUNT]) {
```
Comment (markshannon) [inline Python/optimizer.c:400]: This is undefined behavior in C. Although it will mostly work in practice.

**Comment by markshannon** [comment]
Sorry for not reviewing this earlier.

I'd rather not use `tp_is_gc`, as it adds quite a bit of overhead to an already slow process.

As a short term fix, we could change `tp_traverse` to return immediately if the executor is immortal.
Longer term, we should fix this in the GC. The GC shouldn't be visiting immortal objects.

File: Python/optimizer.c (hunk 2 above)
Comment (gvanrossum) [reply Python/optimizer.c:400]: So I take it this should work?
```suggestion
    if (_Py_IsImmortal(o)) {
```

Comment (markshannon) [reply Python/optimizer.c:400]: Yes

Comment (gvanrossum) [reply Python/optimizer.c:400]: Maybe add a similar thing to the top of `executor_traverse()`? (Although, really, cold exits don't have exits, so their `exit_count` should always be zero, right? So why bother.

Anyway, see gh-118182 for a fix.

**Review by szef1233 — APPROVED** [review]

**Pull request merged**